### PR TITLE
Laboratory work #4, Anastasia Gubarenko - 21FPL2

### PR DESCRIPTION
**Pull request by cybern00dle**

### CODE REVIEW

File: lab_4_summarization_textrank/main.py [hunk 1]
```diff
@@ -313,4 +480,11 @@ def reply(self, query: str, n_summaries: int = 3) -> str:
...
         """
-        pass
+        if not query or not isinstance(query, str):
+            raise IncorrectQueryError('Incorrect query. Use string as input.')
+        type_check(n_summaries, int)
+        if len(self._knowledge_database) < n_summaries:
+            raise ValueError
+        keywords = self._text_preprocessor.preprocess_text(query)
+        texts = self._find_texts_close_to_keywords(keywords, n_summaries)
+        return 'Ответ:\n' + '\n\n'.join(self._knowledge_database[text]['summary'] for text in texts)
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/main.py:490]: use fstring

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py (hunk 1 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py]: done

File: lab_4_summarization_textrank/main.py [hunk 2]
```diff
@@ -2,77 +2,130 @@
...
 from lab_3_keywords_textrank.main import TextEncoder, \
-    TextPreprocessor
+    TextPreprocessor, TFIDFAdapter
 
 PreprocessedSentence = tuple[str, ...]
 EncodedSentence = tuple[int, ...]
 
 
+def type_check(check_object: Any, object_type: Union[Type, tuple[Type, ...]], token_type: Any = None,
+               value_type: Any = None, can_be_empty: bool = True) -> None:
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/main.py:16]: value_type: Type

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py (hunk 2 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py]: done

File: lab_4_summarization_textrank/main.py [hunk 3]
```diff
@@ -2,77 +2,130 @@
...
 
 from lab_3_keywords_textrank.main import TextEncoder, \
-    TextPreprocessor
+    TextPreprocessor, TFIDFAdapter
 
 PreprocessedSentence = tuple[str, ...]
 EncodedSentence = tuple[int, ...]
 
 
+def type_check(check_object: Any, object_type: Union[Type, tuple[Type, ...]], token_type: Optional[Type] = None,
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:15]: decouple

File: lab_4_summarization_textrank/main.py [hunk 4]
```diff
@@ -2,77 +2,130 @@
...
+    :param token_type: expected type of its tokens (or keys if the object is a dictionary)
+    :param value_type: expected type of its values (for dictionaries)
+    :param can_be_empty: True if the object can be empty, False otherwise
+    :return: None
+    """
+    if not can_be_empty and not check_object:
+        raise ValueError
+    if not isinstance(check_object, object_type):
+        raise ValueError
+    if object_type == int and isinstance(check_object, bool):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:30]: demonstrate

File: lab_4_summarization_textrank/main.py [hunk 5]
```diff
@@ -2,77 +2,130 @@
...
+        raise ValueError
+    if isinstance(check_object, dict) and token_type and value_type:
+        for token, value in check_object.items():
+            if not isinstance(token, token_type) or not isinstance(value, value_type):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:34]: ?

File: lab_4_summarization_textrank/main.py [hunk 6]
```diff
@@ -120,23 +184,35 @@ def __init__(self) -> None:
...
 
     def _learn_indices(self, tokens: tuple[str, ...]) -> None:
         """
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        type_check(tokens, tuple, str)
+        new_tokens = (token for token in tokens if token not in self._word2id)
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:197]: type(new_tokens)?

File: lab_4_summarization_textrank/main.py [hunk 7]
```diff
@@ -146,7 +222,13 @@ def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[lis
     :param other_sequence: a sequence of items
     :return: similarity score
     """
-    pass
+    type_check(sequence, (list, tuple))
+    type_check(other_sequence, (list, tuple))
+    if not sequence or not other_sequence:
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:227]: .

File: lab_4_summarization_textrank/main.py [hunk 8]
```diff
@@ -146,7 +222,13 @@ def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[lis
...
+        return 0.0
+    ab_common = set(sequence) & set(other_sequence)
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:229]: .

File: lab_4_summarization_textrank/main.py [hunk 9]
```diff
@@ -193,15 +297,25 @@ def get_similarity_score(self, sentence: Sentence, other_sentence: Sentence) ->
         :param other_sentence
         :return: the similarity score
         """
-        pass
+        type_check(sentence, Sentence)
+        type_check(other_sentence, Sentence)
+        if sentence not in self._vertices or other_sentence not in self._vertices:
+            raise ValueError
+        idx1 = self._vertices.index(sentence)
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:304]: think about problems

Comment (demid5111) [inline lab_4_summarization_textrank/main.py:304]: think about problems

File: lab_4_summarization_textrank/main.py [hunk 10]
```diff
@@ -229,7 +348,12 @@ def update_vertex_score(
         :param scores: current vertices scores
         :return:
         """
-        pass
+        type_check(vertex, Sentence)
+        type_check(incidental_vertices, list, Sentence)
+        type_check(scores, dict, Sentence, float)
+        summa = sum((1 / self._graph.calculate_inout_score(inc_vertex)) * scores[inc_vertex]
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:354]: choose another name

File: lab_4_summarization_textrank/main.py [hunk 11]
```diff
@@ -304,7 +466,14 @@ def _find_texts_close_to_keywords(self, keywords: tuple[str, ...], n_texts: int)
         :param n_texts: number of texts to find
         :return: the texts' ids
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:467]: train of what happened during demonstration

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py (hunk 6 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py]: generator, fixed

File: lab_4_summarization_textrank/main.py (hunk 3 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py]: done

File: lab_4_summarization_textrank/main.py (hunk 5 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py:49]: done, bool checks added

File: lab_4_summarization_textrank/main.py (hunk 10 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py]: done, fixed

File: lab_4_summarization_textrank/main.py (hunk 11 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py:488]: of TextRankSummarizer

File: lab_4_summarization_textrank/main.py (hunk 4 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py]: will demonstrate

File: lab_4_summarization_textrank/main.py (hunk 9 above)
Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py:325]: will coment

Comment (cybern00dle) [reply lab_4_summarization_textrank/main.py:325]: will comment

**Pull request closed without merging**